src/main/java/tigase/xmpp/impl/MessageCarbons.java

```diff
@@ -138,6 +152,9 @@ public void setMsgCarbonPaths(String[] matcherStrs) {
 			}
 		}
 		msgCarbonPaths = matchers.toArray(new ElementMatcher[0]);
+		usesDefCarbonRules = Arrays.equals(
+				Arrays.stream(DEF_MSG_CARBON_PATHS).map(matcher -> matcher.toString()).toArray(String[]::new),
+				matcherStrs);
```

Comment:
1) Using `Arrays.equals()` compares also order so if the paths are the same but in different order in it still may result in `false` being the result?
2) If someone would want to extend the default set and _add_ some more paths then it should still yield `true` but in this case it wouldn't.

Maybe using `Collections.containsAll()` against default/required set would be better here?

Reply:
Actually, this was intended.

Using `containsAll()` will work if someone would prepend our rules with `ElementMatcher` which would return `false`. That would break rules and server would still report `urn:xmpp:carbons:rules:0` while it should not, according to the XEP:

> A server implementation can choose to advertise full support of all the rules in §6.1 by including the "urn:xmpp:carbons:rules:0" feature in its service discovery information. If that feature is advertised, the rules above must be treated as REQUIRED and not merely as RECOMMENDED.
> 
> Accordingly, if this feature is advertised, a client MAY rely on the server supporting this exact set of rules.